lib/Modules/HyperVCIHooks/HyperVCIHooks.psm1

```diff
@@ -0,0 +1,1482 @@
+        Start-ExternalCommand { git -C $projectDir clean -x -f -d -q } -ErrorMessage "Failed to git clean"
+    }
+
+    echo "Before doing git checkout:"
+    echo "Git branch output:"
+    Start-ExternalCommand { git -C $projectDir branch } -ErrorMessage "Failed to show git branch."
+    echo "Git log output:"
+    Start-ExternalCommand { git -C $projectDir log -10 --pretty=format:"%h - %an, %ae, %ar : %s" } `
+        -ErrorMessage "Failed to show git log."
+
+    $ret = echo "$ZuulRef" | Where-Object { $_ -match "^refs/tags/" }
+    if ($ret) {
+        Start-ExternalCommand {
+                git -C $projectDir fetch --tags "$ZuulUrl/$ZuulProject"
+                git -C $projectDir checkout $ZuulRef
+                git -C $projectDir reset --hard $ZuulRef
+            } `
+                    -ErrorMessage "Failed to fetch tags from: $ZuulUrl/$ZuulProject"
+    }
+    elseif (!$ZuulNewrev) {
+        Start-ExternalCommand {
+                git -C $projectDir fetch "$ZuulUrl/$ZuulProject" $ZuulRef
+                git -C $projectDir checkout FETCH_HEAD
+                git -C $projectDir reset --hard FETCH_HEAD
+            } `
+                    -ErrorMessage "Failed to fetch: $ZuulUrl/$ZuulProject $ZuulRef"
+    }
+    else {
+        Start-ExternalCommand {
+                git -C $projectDir checkout $ZuulNewrev
+                git -C $projectDir reset --hard $ZuulNewrev
+            } `
+                    -ErrorMessage "Failed to checkout $ZuulNewrev"
+    }
+
+    try {
+        Start-ExternalCommand { git -C $projectDir clean -x -f -d -q } -ErrorMessage "Failed to git clean"
+    } catch {
+        sleep 1
+        Start-ExternalCommand { git -C $projectDir clean -x -f -d -q } -ErrorMessage "Failed to git clean"
+    }
+
+    if (Test-Path $gitmodulesDir) {
+        Start-ExternalCommand { 
+                git -C $projectDir submodule init
+                git -C $projectDir submodule sync
+                git -C $projectDir submodule update --init
+            } `
+                    -ErrorMessage "Failed to init submodule"
+    }
+
+    echo "Final result:"
+    echo "Git branch output:"
+    Start-ExternalCommand { git -C $projectDir branch } -ErrorMessage "Failed to show git branch."
+    echo "Git log output:"
+    Start-ExternalCommand { git -C $projectDir log -10 --pretty=format:"%h - %an, %ae, %ar : %s" } `
+        -ErrorMessage "Failed to show git log."
+
+}
+
+
+function Install-Projects ($buildFor) {
+    # Shared for all projects
+    $projectDir = $buildFor -replace 'openstack/'
+    Write-JujuLog "Installing $projectDir"
+    $openstackBuild = Join-Path $BUILD_DIR "openstack"
+    Start-ExecuteWithRetry {
+        Install-OpenStackProjectFromRepo "$openstackBuild\$projectDir"
+    }
+    
+    # Individual projects
+    # Nova
+    if ($buildFor -eq "openstack/nova") {
+        Invoke-BuildNova
+    }
+    # Networking-Hyperv
+    if ($buildFor -eq "openstack/networking-hyperv") {
+        Invoke-BuildNeutron
```

Comment:
ok...this was a poor choice of function parameter. Will change it into "function Install-Projects ($project)". 
This function is called when installing any openstack service, not just the gerrit one. So the param for this function will never be empty. 
Second:
Invoke-BuildNeutron is only called when networking-hyperv is installed because it just checks to see if 'neutron-hyperv-agent.exe' exists on the system. When OVS is the networking driver, afaik, there is no need to check for 'neutron-hyperv-agent.exe' because it does not exist and will not be used.
For 'openstack/compute-hyperv' i was under the impression that the config files that are copied in the invoke-buildnova function are the same as the 'openstack/nova' ones, so no need to copy them one more time. But i feel that i may be wrong on this one!